### DISCUSSION TITLE
peer to peer instead of client to server

### DISCUSSION THREAD

**Discussion by jihiggs** [input-leap/input-leap#1599] [Ideas]
mouse without borders does this, this means you can use any keyboard and mouse to control any computer that has synced with the other computers running mouse without borders.  this allows me to sit in my recliner with my laptop and move my mouse to one of the other computers on my desk to do something, while at the same time allow me to use the keyboard and mouse from any of those computers to control any of the others.

**Comment by shymega** [comment]
This would require a huge effort on our part, and a total overhaul of the design of Input Leap, plus a breaking change to the protocol. Can't see it being done, unfeasible with such a small team.

**Reply by jihiggs** [reply]
oh well, just a thought

**Reply by Cimbali** [reply]
Currently you have to swap the configuration on both machines (on server, change to client; on client, change to server).

- Would a way to trigger that from the client (only) solve your use case @jihiggs ?
- Do you think that’s still a lot of work @shymega ?

It’d only be a message from client to server, triggering reconfiguration, then back to usual work, so there’s never a “any machine can be the server” sort of situation. I’d be willing to look into that, if it seems doable.

**Reply by jihiggs** [reply]
that would likely require restarting the service on each machine as well.  not sure how possible that is.  the program could be configured to watch the config file for changes.  might not be very efficient.

**Reply by Cimbali** [reply]
No I meant as a functionality of input-leap, that would modify the active configuration, not some external script that modifies config files or something like that. It would likely require restarting all connections, pretty much what the `Reload` button already does.

**Reply by shymega** [reply]
> Currently you have to swap the configuration on both machines (on server, change to client; on client, change to server).
> 
> * Would a way to trigger that from the client (only) solve your use case @jihiggs ?
> * Do you think that’s still a lot of work @shymega ?
> 
> It’d only be a message from client to server, triggering reconfiguration, then back to usual work, so there’s never a “any machine can be the server” sort of situation. I’d be willing to look into that, if it seems doable.

Yes, we could implement a 'reload from specified file' command over the protocol, but it'd still be a change to the protocol and would break compatibility with Barrier - if that would affect people.

**Comment by miezebieze** [comment]
has anyone tried to run a server on every peer on a different port each and on every peer a client per server?
i'm gonna try that now. wish me luck...

it works -- kind of.

of course my first idea was using the same layout on all (two) peers.
in this case the pointer can be moved once it's on the remote screen, but after moving it to the one the mouse/touchpad is connected to, it cannot get back to the remote screen.

i also tried to use a different layout (mirrored) on the second server.
this works for a moment and connects the outer edges of the screeen before the pointer disappears and X becomes unresponsive on both peers. (it didn't lag out and i was able to switch to another tty and kill the program.)

**Reply by shymega** [reply]
> has anyone tried to run a server on every peer on a different port each and on every peer a client per server? i'm gonna try that now. wish me luck...
> 
> it works -- kind of.
> 
> of course my first idea was using the same layout on all (two) peers. in this case the pointer can be moved once it's on the remote screen, but after moving it to the one the mouse/touchpad is connected to, it cannot get back to the remote screen.
> 
> i also tried to use a different layout (mirrored) on the second server. this works for a moment and connects the outer edges of the screeen before the pointer disappears and X becomes unresponsive on both peers. (it didn't lag out and i was able to switch to another tty and kill the program.)

Good to know.

**Comment by miezebieze** [comment]
i just had an idea how peer to peer could be achieved. well, it's not "true" p2p, but should feel like it.

the input-leap client needs to grab all input events and send the non input-leap to the server (before the event does anything else).
the server sends the events to whatever client is in focus. (even if that means sending the event right back where it came from.)

it should be completely backwards compatible as well. outdated clients would still be able to receive events.

**Reply by shymega** [reply]
Hm. Maybe. I can't make any promises. I'm not even sure if Povilas (@p12tic) would want a change to the protocol. I remember a discussion on IRC about it recently.
